src/Haskell/Ide/Engine/Transport/LspStdio.hs

```diff
@@ -97,33 +105,50 @@ run
   -> IO Int
 run dispatcherProc cin _origDir plugins captureFp = flip E.catches handlers $ do
 
-  rin <- atomically newTChan :: IO (TChan ReactorInput)
+  rin        <- atomically newTChan :: IO (TChan ReactorInput)
   commandIds <- allLspCmdIds plugins
 
   let dp lf = do
         cancelTVar  <- atomically $ newTVar S.empty
         wipTVar     <- atomically $ newTVar S.empty
         versionTVar <- atomically $ newTVar Map.empty
+        diagIn      <- atomically newTChan
         let dEnv = DispatcherEnv
               { cancelReqsTVar = cancelTVar
               , wipReqsTVar    = wipTVar
               , docVersionTVar = versionTVar
               }
-        let reactorFunc = runReactor lf dEnv cin diagnosticProviders hps sps
-              $ reactor rin
+        let react = runReactor lf dEnv cin diagIn diagnosticProviders hps sps
+        let reactorFunc = react $ reactor rin
             caps = Core.clientCapabilities lf
 
         let errorHandler :: ErrorHandler
             errorHandler lid code e =
               Core.sendErrorResponseS (Core.sendFunc lf) (J.responseId lid) code e
             callbackHandler :: CallbackHandler R
-            callbackHandler f x = runReactor lf dEnv cin diagnosticProviders hps sps $ f x
+            callbackHandler f x = react $ f x
+
+        -- This is the callback the debouncer executes at the end of the timeout,
+        -- it executes the diagnostics for the most recent request.
+        let dispatchDiagnostics :: Maybe (MostRecent DiagnosticsRequest) -> R ()
+            dispatchDiagnostics Nothing = pure ()
+            dispatchDiagnostics (Just (MostRecent req)) = performDiagnostics req
+
+        -- Debounces messages published to the diagnostics channel.
+        let diagnosticsQueue tr = forever $ do
+              inval <- liftIO $ atomically $ readTChan diagIn
+              Debounce.send tr (Just $ MostRecent inval)
 
+        tr <- Debounce.new
+          (Debounce.forMonoid $ react . dispatchDiagnostics)
+          (Debounce.def { Debounce.delay = 350000, Debounce.alwaysResetTimer = True })
```

Comment:
This number means 0.35sec, if I understand it correctly.

Reply:
that's correct